### PR TITLE
[inference][trt]set output data type of trt network

### PR DESCRIPTION
**Pull request by zhangjun**
### PR types

Others
### PR changes

Others
### Describe

- Set network output's from programdesc.
- Upgrade the `cast` op to support all data types (bool, int32, int64, half, float).
- Disable `equal` op  with implicit batch. Because TensorRT does not support kEQUAL/kGREATER/kLESS working with implicit batch
- #47304  is similar to this PR， but ```map_origin_outputs_dtype``` in ```tensorrt_subgraph_pass``` is not set correctly.

### CODE REVIEW

File: paddle/fluid/inference/tensorrt/engine.cc [hunk 1]
```diff
@@ -207,16 +207,19 @@ void TensorRTEngine::FreezeNetwork() {
...
-      network()->getOutput(i)->setType(nvinfer1::DataType::kFLOAT);
+  for (int i = 0; i < network()->getNbInputs(); ++i) {
+    auto *input = network()->getInput(i);
+    switch (input->getType()) {
+      case nvinfer1::DataType::kINT32:
+      case nvinfer1::DataType::kBOOL:
+      case nvinfer1::DataType::kHALF:
+        break;
+      case nvinfer1::DataType::kFLOAT:
+      case nvinfer1::DataType::kINT8:
```
Comment (jiweibo) [inline paddle/fluid/inference/tensorrt/engine.cc]: int8的输入，要映射trt的float吗？

File: paddle/fluid/inference/tensorrt/engine.cc [hunk 2]
```diff
@@ -422,6 +425,25 @@ void TensorRTEngine::DeclareOutput(const std::string &name) {
                         name));
   network()->markOutput(*output);
 }
+
+void TensorRTEngine::DeclareOutput(const std::string &name,
+                                   nvinfer1::DataType dtype) {
+  auto *output = TensorRTEngine::GetITensor(name);
```
Comment (jiweibo) [inline paddle/fluid/inference/tensorrt/engine.cc:416]: 复用上面`DeclareOutput(const std::string &name) `的代码

File: paddle/fluid/inference/tensorrt/engine.cc [hunk 3]
```diff
@@ -207,16 +207,19 @@ void TensorRTEngine::FreezeNetwork() {
...
+  for (int i = 0; i < network()->getNbInputs(); ++i) {
+    auto *input = network()->getInput(i);
+    switch (input->getType()) {
+      case nvinfer1::DataType::kINT32:
+      case nvinfer1::DataType::kBOOL:
+      case nvinfer1::DataType::kHALF:
+        break;
+      case nvinfer1::DataType::kFLOAT:
+      case nvinfer1::DataType::kINT8:
+        input->setType(nvinfer1::DataType::kFLOAT);
```
Comment (Wangzheee) [inline paddle/fluid/inference/tensorrt/engine.cc]: 这里为什么要设置 input 的type呢？DeclareInput 会在 op_converter.h里 遍历所有的input根据type进行设置

File: paddle/fluid/inference/tensorrt/engine.cc (hunk 2 above)
Comment (zhangjun) [reply paddle/fluid/inference/tensorrt/engine.cc:416]: > 复用上面`DeclareOutput(const std::string &name) `的代码

对 这里可以复用上面函数，下个PR改掉

File: paddle/fluid/inference/tensorrt/engine.cc (hunk 3 above)
Comment (zhangjun) [reply paddle/fluid/inference/tensorrt/engine.cc]: 已remove

**Review by zhoutianzi666 — APPROVED** [review]
LGTM

**Review by Wangzheee — APPROVED** [review]
LGTM

**Pull request merged**